python/tskit/tables.py

```diff
@@ -2781,3 +2782,204 @@ def union(
+            edge_choice = rng.integers(0, np.arange(1, len(child_ids) * 2 - 1, 2))
+            tmp_new_node_lab = [parent_node_id] + new_nodes_by_time_desc
+            assert len(edge_choice) == len(child_ids) - 1
+            for node_lab, child_id, target_edge_id in zip(
+                tmp_new_node_lab, child_ids[1:], edge_choice
+            ):
+                target_edge = edges[target_edge_id]
+                # Insert in the right place, to keep edges in parent time order
+                edges.insert(target_edge_id, [child_id, node_lab])
+                edges.insert(target_edge_id, [target_edge[0], node_lab])
+                target_edge[0] = node_lab
+            top_edge = edges.pop()  # remove the edge above the top node
+            assert top_edge[1] is None
+
+            # Re-map the internal nodes IDs so they are used in time order
+            real_node = iter(new_nodes_by_time_desc)
+            node_map = {c: c for c in child_ids}
+            node_map[edges[-1][1]] = parent_node_id  # last edge == oldest parent
+            for e in reversed(edges):
+                # Reversing along the edges, parents are in inverse time order
+                for idx in (1, 0):  # look at parent (1) then child (0)
+                    if e[idx] not in node_map:
+                        node_map[e[idx]] = next(real_node)
+                    e[idx] = node_map[e[idx]]
+            assert len(node_map) == len(new_nodes_by_time_desc) + len(child_ids) + 1
+            return edges
+
+        edge_table = self.edges
+        node_table = self.nodes
+        # Store existing left, so we can change it if the edge is split
+        existing_edges_left = edge_table.left
+        # Keep other edge arrays etc. for fast read access
+        existing_edges_right = edge_table.right
+        existing_edges_parent = edge_table.parent
+        existing_edges_child = edge_table.child
+        existing_node_time = node_table.time
+
+        # We can save a lot of effort if we don't need to check the time of mutations
+        # We definitely don't need to check on the first iteration, a
+        check_mutations = np.any(
+            np.logical_not(is_unknown_time_array(self.mutations.time))
+        )
+        ts = self.tree_sequence()  # Only needed to check mutations
+        tree_iter = ts.trees()  # ditto
+
+        edge_table.clear()
+
+        edges_from_node = collections.defaultdict(set)  # Active descendant edge ids
+        nodes_changed = set()
+
+        for interval, e_out, e_in in ts.edge_diffs(include_terminal=True):
+            pos = interval[0]
+            prev_tree = None if pos == 0 else next(tree_iter)
+
+            for edge in itertools.chain(e_out, e_in):
+                if edge.parent != tskit.NULL:
+                    nodes_changed.add(edge.parent)
+
+            if check_mutations and prev_tree is not None:
+                # This is grim. There must be a more efficient way.
```

Comment:
@brianzhang01 - can you think of a nicer way to find the oldest mutation above a node?

Reply:
I just thought - we can probably do this by pre-sorting a copy of the mutations - sorting first by node, then by time (if there are no mutation times, then we don't need to do this anyway). The slight problem is that we might have mutations that are oldest above a node, but not on the current tree. We should be able to mask those out, however, by taking the list of mutations for this node and restricting to only the sites in this tree.